### PR TITLE
Various encodings conversion fixes

### PR DESCRIPTION
**Pull request by b4n**
* Fix detection of some non-UTF-8 files with a encoding stanza in them (e.g. `coding: iso-8859-2\nĄ őŕ B`)
* Fix out-of-bounds reads loading an empty UTF-16 buffer with BOM (yeah, really)
* Fix **silently** truncating at the first NUL in some case (part of #3700)
* Fix handling of non-UTF-8 data that start with a UTF-8 BOM

Note: this still doesn't load data after the first NUL byte, but at least now it will always properly warn as we supposedly were already doing.

### CODE REVIEW

**Comment by elextr** [comment]
LGBI, I didn't inspect the test code more than a (very) quick skim.  Compiles ok, but encoding tests not made because not in `tests/meson.build` and being an ASCII locale I have no other encodings, so untested.  

The whole encodings thing is still a mess, but its now looks a more correct mess :-)

PS isn't the milestone a bit specific? :wink:

**Comment by b4n** [comment]
> Compiles ok, but encoding tests not made because not in `tests/meson.build`

Oopsie, fixed.  Aaah, that joy of having multiple build systems to maintain is back, I almost forgot it after Waf bit the dust…

**Comment by elextr** [comment]
> that joy of having multiple build systems to maintain is back

Feel free to remove autotools :wink:

Anyway, tests seem to pass, although with 10 billion[^1] `Geany-INFO` messages.

Several of these messages appear:

```
# Geany-INFO: Couldn't convert from ISO-8859-8-I to UTF-8 (Conversion from character set “ISO-8859-8-I” to “UTF-8” is not supported).
```

What that points out is that Geany has a fixed list of conversions, which are exposed to user selection, but not all systems support all conversions, as you might guess my `iconv -l` does not contain `ISO-8859-8-I`.  If I try to open an ASCII file (which is valid for all 8859s) with `ISO-8859-8-I` I get `The file xxxx is not valid ISO-8859-8-I` which is wrong, does `g_convert()` return a better error we could show (like the one from the test)?

[^1]: slight exaggeration for effect

**Comment by b4n** [comment]
> Several of these messages appear:
> 
> ```
> # Geany-INFO: Couldn't convert from ISO-8859-8-I to UTF-8 (Conversion from character set “ISO-8859-8-I” to “UTF-8” is not supported).
> ```
>
> What that points out is that Geany has a fixed list of conversions, which are exposed to user selection, but not all systems support all conversions, as you might guess my `iconv -l` does not contain `ISO-8859-8-I`. If I try to open an ASCII file (which is valid for all 8859s) with `ISO-8859-8-I` I get `The file xxxx is not valid ISO-8859-8-I` which is wrong[2](#user-content-fn-2-75b2eae86374246fedd5de48e30f3431), does `g_convert()` return a better error we could show (like the one from the test)?

`g_convert()` reports the message you see in the info message inside the parentheses.  So yes, we could display it, it's a *Mere Matter of Programming™*, to report that error all the way up to the caller that will show the error.

As of [ISO-8859-8-I](https://en.wikipedia.org/wiki/ISO/IEC_8859-8) it seems understandable not to be supported, as IIUC it's the same thing that should just be *rendered* differently (or not, which makes matter worse).

Anyhow, yes, this could be improved, and ideally we'd probably not display the encoding in the list if it's not supported by the library we use (iconv), as it makes little sense then.  But I'd argue this should probably be a separate improvement :)

**Comment by elextr** [comment]
> Anyhow, yes, this could be improved, and ideally we'd probably not display the encoding in the list if it's not supported by the library we use (iconv), as it makes little sense then.

Would be nice but IIRC last time we discussed this there was no easy way of finding what was supported by `g_convert()`, and does that only use iconv, or does it use something else on Windows and Macos?  And even if its iconv all the way, there is no programmatic way of determining what it supports.  So probably it would involve shell testing `iconv -l` at install time and adapting the menu. Too hard.

> So yes, we could display it, it's a Mere Matter of Programming™, to report that error all the way up to the caller that will show the error.

A perfectly good thing to put in a PR called "Various encodings conversion fixes" he says hopefully?  Maybe the lowest level function could display the message if it was `force` otherwise don't worry about it because its just Geany trying all encodings it knows again ;-)

**Comment by b4n** [comment]
> > Anyhow, yes, this could be improved, and ideally we'd probably not display the encoding in the list if it's not supported by the library we use (iconv), as it makes little sense then.
> 
> Would be nice but IIRC last time we discussed this there was no easy way of finding what was supported by `g_convert()`, and does that only use iconv, or does it use something else on Windows and Macos?

It's (almost, but is for all practical purposes) documented as using `g_iconv()`, which is documented as using `iconv()` or libiconv as fallback.

> And even if its iconv all the way, there is no programmatic way of determining what it supports.

Oops, I didn't know it was impossible so I did it in the meantime :smile:  (not pushed here, but I have something locally)
More seriously though, no there don't seem to be any way of querying the list of encodings it supports, but there is a way to query whether a particular combination is supported or not: just use `g_iconv_open(A, B)` and it'll error out with `EINVAL` if the conversion from B to A is not supported.

> So probably it would involve shell testing `iconv -l` at install time and adapting the menu. Too hard.

Yeah, no, that's not gonna happen I don't think.

> > So yes, we could display it, it's a Mere Matter of Programming™, to report that error all the way up to the caller that will show the error.
> 
> A perfectly good thing to put in a PR called "Various encodings conversion fixes" he says hopefully?

Fair enough :smile:   I should stop those fairly generic names even when I ended up fixing a various things on the way, I end up getting sidetracked forever :upside_down_face: 

> Maybe the lowest level function could display the message if it was `force` otherwise don't worry about it because its just Geany trying all encodings it knows again ;-)

Meh, that's a bit to ~~Geanyic~~as-hoc, but the error could definitely be propagated.

**Comment by elextr** [comment]
> It's (almost, but is for all practical purposes) documented as using g_iconv(), which is documented as using iconv() or libiconv as fallback.

Ok, the [code-ocumentation](https://github.com/GNOME/glib/blob/main/glib/gconvert.c) shows its iconv all the way, not a #ifdef windows in sight :-)

>  just use g_iconv_open(A, B) and it'll error out with EINVAL if the conversion from B to A is not supported.

So we can use that in `init_encodings()` to remove unsupported ones from the list.

It won't add any not supported by whoever made the list in `encodings.c`, but hopefully in this time of Unicode that won't be many.

**Comment by b4n** [comment]
> A perfectly good thing to put in a PR called "Various encodings conversion fixes" he says hopefully? Maybe the lowest level function could display the message if it was `force` otherwise don't worry about it because its just Geany trying all encodings it knows again ;-)

Would you really like me to add this here, in addition to the already non-trivial changes, or should this be a follow-up?

I've got a few more cosmetic changes, including propagating the error better as you wished, as well as not showing knowingly unsupported encodings in the UI and removing a unnecessary control in the preferences regarding file encodings.  Should I shove everything here because it's encoding-related, or should we keep the "fix encodings conversion" separate from the more cosmetic stuff?

> It won't add any not supported by whoever made the list in `encodings.c`, but hopefully in this time of Unicode that won't be many.

:crossed_fingers:

**Comment by elextr** [comment]
> including propagating the error better as you wished, as well as not showing knowingly unsupported encodings in the UI

In the end thats all I am suggesting adding now.

> removing a unnecessary control in the preferences regarding file encodings. 

Which one?  I only found encoding of new files and use fixed encoding both of which seem useful?

**Comment by elextr** [comment]
> or should we keep the "fix encodings conversion" separate from the more cosmetic stuff?

Well, ATM its just tests and fixes AFAICT, not sure what you see as "cosmetic"?

PS sure its several fixes, but they are all just part of making encoding work correctly, or as correctly as it can be made to work without rewriting.

**Comment by b4n** [comment]
> In the end thats all I am suggesting adding now.

Done (plus a couple of weird loop I just couldn't let be).  Let me know how I should update all strings :wink: 

> Which one? I only found "encoding of new files" and "use fixed encoding" both of which seem useful?

*Use fixed encoding when opening non-Unicode files*.  It's useful, but we have code for adding a *Detect from file* entry directly in the combo box, as used in the file Open dialog.  Using this there as well instead of having a separate checkbox strikes me as an improvement: simpler UI with less widgets and less code.

> Well, ATM its just tests and fixes AFAICT, not sure what you see as "cosmetic"?

Cosmetic is the other this I have for the UI, and I count the more proper error message as cosmetic compared to the other fixes.

**Comment by elextr** [comment]
> Use fixed encoding when opening non-Unicode files. It's useful, but we have code for adding a Detect from file entry directly in the combo box, as used in the file Open dialog. Using this there as well instead of having a separate checkbox strikes me as an improvement: simpler UI with less widgets and less code.

Well, IIUC the "use fixed encoding" is intended to save the user from having to switch the forced encoding on the open dialog from Unicode to fixed all the time, like on a Windows machine with a non-unicode default code page, but the user is also editing Unicode HTML.  So if my understanding is correct its a useful convenience, not essential.  But for sure moving it to the open dialog will make it more discoverable.

**Comment by b4n** [comment]
@elextr it's already in the open dialog, and I'm not gonna move anything.  I'm just gonna merge *Use fixed encoding when opening non-Unicode files* checkbox and *Default encoding (existing non-Unicode files)* combo, only keeping the combo and having an additional *Detect from file* item in it which takes over the current *Use fixed encoding when opening non-Unicode files* checkbox role.

**Comment by elextr** [comment]
>  I count the more proper error message as cosmetic compared to the other fixes.

And as you may have guessed I see them as fixes, so of course I say all in together :wink:

Will check the strings as soon as I find my purple editing pencil.

**Comment by elextr** [comment]
> @elextr it's already in the open dialog, and I'm not gonna move anything. I'm just gonna merge Use fixed encoding when opening non-Unicode files checkbox and Default encoding (existing non-Unicode files) combo, only keeping the combo and having an additional Detect from file item in it which takes over the current Use fixed encoding when opening non-Unicode files checkbox role.

Ok, that one is cosmetic, removing a checkbox and adding a do nothing combo setting instead.

But I disagree that its the same as the encoding selection in the open dialog, the open dialog sets encoding to use for _all_ files, the one in prefs sets it for _non-unicode_ files IIUC.  I didn't check the code, so if they in fact work the same then there is a bug because one of the functions is missing, or it was just described wrong and then the labels need fixing to say the right thing.  Don't forget the manual if changing UI.

**Comment by b4n** [comment]
> But I disagree that its the same as the encoding selection in the open dialog

I'm not saying it's the same, just that we have the code for generating the combo box with the additional item as used in that dialog, so I don't even have to add this.
All this commenting is for the future PR with it, no point in discussing thin air, it's confusing for no reason ;)

**Comment by elextr** [comment]
> All this commenting is for the future PR with it, no point in discussing thin air, it's confusing for no reason ;)

Ok, waiting for separate concrete :smile:

**Comment by b4n** [comment]
@elextr I'll pour it as soon as this gets merged ;) (unless I find out there's no dependency between the two, in case I can get those to cure in parallel)

**Comment by eht16** [comment]
There are actually some encodings which seem not support on Windows:
```
15:31:32.356445: Geany INFO		: Encoding ISO-8859-10 is not supported by the system
15:31:32.356445: Geany INFO		: Encoding ISO-8859-14 is not supported by the system
15:31:32.356445: Geany INFO		: Encoding ISO-8859-16 is not supported by the system
15:31:32.356445: Geany INFO		: Encoding UTF-7 is not supported by the system
15:31:32.356445: Geany INFO		: Encoding ARMSCII-8 is not supported by the system
15:31:32.356445: Geany INFO		: Encoding EUC-TW is not supported by the system
15:31:32.356445: Geany INFO		: Encoding GEORGIAN-ACADEMY is not supported by the system
15:31:32.356445: Geany INFO		: Encoding HZ is not supported by the system
15:31:32.356445: Geany INFO		: Encoding ISO-IR-111 is not supported by the system
15:31:32.356445: Geany INFO		: Encoding TCVN is not supported by the system
15:31:32.356445: Geany INFO		: Encoding TIS-620 is not supported by the system
15:31:32.356445: Geany INFO		: Encoding VISCII is not supported by the syste
```

For the rest, I did some basic testing on Linux and works as expected and by inspection it looks (almost) fine.
Even if there might be subtile bugs on edge cases, I would merge this in master already to get broader testing of the changes.

File: src/encodings.c [hunk 1]
```diff
@@ -639,18 +625,22 @@ gchar *encodings_convert_to_utf8_from_charset(const gchar *buffer, gssize size,
...
 			geany_debug("Couldn't convert from %s to UTF-8 (%s).", charset, conv_error->message);
-			g_error_free(conv_error);
+			g_propagate_error(error, conv_error);
 			conv_error = NULL;
 		}
 		else
+		{
 			geany_debug("Couldn't convert from %s to UTF-8.", charset);
+			g_set_error(error, G_CONVERT_ERROR, G_CONVERT_ERROR_ILLEGAL_SEQUENCE,
+					_("Data contains NULs"));
```
Comment (eht16) [inline src/encodings.c:642]: Do we always know for sure the error here is exclusively NUL bytes?

**Comment by b4n** [comment]
@elextr Tell me if there are still strings to be updated or better error to report, I can make another follow-up :wink:

**Comment by elextr** [comment]
@b4n the only one that I would question is the one that @eht16 had questioned, so I was waiting until that was decided. :-P

"Data contains NULs" is referring to the result of the conversion, but the user doesn't know that, they only have the input file and it may not contain NULs, eg it might have an overlong encoding of a nul.  

So "Result of conversion to UTF-8 buffer contains NULs" might be better.

**Pull request merged**